uFmt macros don't fully-qualify Ok/Err enum variants

Hi,

I have a small problem with using uFmt (which I otherwise love :)).  Namely, when the `uwrite!`/`uwriteln!` macros are expanded, they don't fully qualify their references to core::result::Result::Ok & core::result::Result::Err.

That means if somewhere else in your code you `use` an alternative implementation of Result, calls to uwrite! start exploding.  Minimum representative example:

```
use ufmt::{derive::uDebug, uwrite};

#[derive(uDebug)]
struct Pair { x: u32, y: u32 }

enum SomeOtherResult<A,B> {
  Ok(A),
  Err(B)
}

use SomeOtherResult::{Ok,Err};

pub fn main() {
  let mut s = String::new();
  let pair = Pair { x: 1, y: 2 };

  uwrite!(s, "{:?}", pair).unwrap();
  assert_eq!(s, "Pair { x: 1, y: 2 }");

  println!("String = {s}");
}
```

This will fail to compile with:

```
error[E0308]: mismatched types
  --> src/main.rs:17:3
   |
17 |   uwrite!(s, "{:?}", pair).unwrap();
   |   ^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `Result`, found enum `SomeOtherResult`
```

The fix here is for the macros to not output 'bare' Ok/Err references, but rather fully qualify them so they can't clash with the calling program's namespace.  (I can probably fix this myself and then put in a pull() request, but want to make sure if it's likely to be accepted first :-).)

Cheers,
Tim


PS. You're probably asking "why would you do that anyway?".  The answer is because the default implementation of Result pulls in Debug in its implementation of unwrap() etc. etc., basically meaning it's unusable in an embedded context.  To allow idiomatically Rust code (with ? and .unwrap() and so on) I provide my own implementation of Result that works sensibly in an embedded system.)

--- a/macros/src/lib.rs
+++ b/macros/src/lib.rs
@@ -140,9 +140,7 @@ pub fn debug(input: TokenStream) -> TokenStream {
 
             let body = if arms.is_empty() {
                 // Debug's implementation uses `::core::intrinsics::unreachable()`
-                quote!(
-                    unsafe { core::unreachable!() }
-                )
+                quote!(unsafe { core::unreachable!() })
             } else {
                 quote!(
                     match self {
@@ -275,7 +273,7 @@ fn write(input: TokenStream, newline: bool) -> TokenStream {
 
             (#formatter).do_as_formatter(|f| {
                 #(#exprs)*
-                Ok(())
+                core::result::Result::Ok(())
             })
         }
     })

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -56,7 +56,7 @@
 //! - Hexadecimal formatting
 //!
 //! Lowercase (`{:x}`), uppercase (`{:X}`), `0x`-prefix (`{:#x}`) and padding (`{:02x}`) are
-//! supported on primitive integer types. 
+//! supported on primitive integer types.
 //!
 //! ```
 //! use ufmt::uwrite;

--- a/tests/gh36.rs
+++ b/tests/gh36.rs
@@ -1,0 +1,38 @@
+//! `uwrite!` works in presence of a third-party `Ok` constructor
+
+mod third_party {
+    #[allow(dead_code)]
+    pub enum Result<A, B> {
+        Ok(A),
+        Err(B),
+    }
+}
+
+use ufmt::{derive::uDebug, uwrite, uwriteln};
+
+#[allow(unused_imports)]
+use crate::third_party::Result::{self, Err, Ok};
+
+#[derive(uDebug)]
+struct Pair {
+    x: u32,
+    y: u32,
+}
+
+#[test]
+fn uwrite() {
+    let mut s = String::new();
+    let pair = Pair { x: 1, y: 2 };
+
+    uwrite!(s, "{:?}", pair).unwrap();
+    assert_eq!(s, "Pair { x: 1, y: 2 }");
+}
+
+#[test]
+fn uwriteln() {
+    let mut s = String::new();
+    let pair = Pair { x: 1, y: 2 };
+
+    uwriteln!(s, "{:?}", pair).unwrap();
+    assert_eq!(s, "Pair { x: 1, y: 2 }\n");
+}

--- a/utils/src/lib.rs
+++ b/utils/src/lib.rs
@@ -9,7 +9,7 @@
 #![deny(warnings)]
 #![no_std]
 
-use core::{convert::Infallible, str, fmt};
+use core::{convert::Infallible, fmt, str};
 
 use heapless::String;
 use ufmt_write::uWrite;
@@ -134,7 +134,6 @@ where
     }
 }
 
-
 /// An adapter struct allowing to use `ufmt` on types which implement `core::fmt::Write`
 ///
 /// For example:
@@ -149,9 +148,14 @@ where
 /// let mut s = String::new();
 /// uwrite!(WriteAdapter(&mut s), "{:?}", fancy_number);
 /// ```
-pub struct WriteAdapter<W>(pub W) where W: fmt::Write;
+pub struct WriteAdapter<W>(pub W)
+where
+    W: fmt::Write;
 
-impl<W> uWrite for WriteAdapter<W> where W: fmt::Write {
+impl<W> uWrite for WriteAdapter<W>
+where
+    W: fmt::Write,
+{
     type Error = fmt::Error;
 
     fn write_char(&mut self, c: char) -> Result<(), Self::Error> {

